### PR TITLE
LibUnicode+LibJS: Add initial implementation of Unicode Segmentation

### PR DESCRIPTION
**Pull request by IdanHo**
This implements the base of version Unicode Segmentation, future improvements include adding locale-sensitive segmentation, as well as adding word-likeness determination for `isWordLike`.

~Reviewers Beware: `LibUnicode: Implement {grapheme, word, sentence} segmentation` are horrific, I would love to hear improvement ideas.~ Improved quite a bit with tryflynn's help

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Intl/SegmenterConstructor.cpp [hunk 1]
```diff
@@ -81,6 +81,7 @@ ThrowCompletionOr<Object*> SegmenterConstructor::construct(FunctionObject& new_t
     auto granularity = TRY(get_option(global_object, *options, vm.names.granularity, Value::Type::String, { "grapheme"sv, "word"sv, "sentence"sv }, "grapheme"sv));
 
     // 13. Set segmenter.[[SegmenterGranularity]] to granularity.
+
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/SegmenterConstructor.cpp:84]: Remove plz :)

File: Userland/Libraries/LibUnicode/CharacterTypes.cpp [hunk 2]
```diff
@@ -357,4 +358,79 @@ bool __attribute__((weak)) code_point_has_grapheme_break_property(u32, GraphemeB
 bool __attribute__((weak)) code_point_has_word_break_property(u32, WordBreakProperty) { return {}; }
 bool __attribute__((weak)) code_point_has_sentence_break_property(u32, SentenceBreakProperty) { return {}; }
 
+Vector<size_t> find_grapheme_segmentation_boundaries([[maybe_unused]] Utf16View const& view)
```
Comment (trflynn89) [inline Userland/Libraries/LibUnicode/CharacterTypes.cpp:361]: What if we use fold expressions to reduce the noise of chaining `code_point_has_grapheme_break_property` calls on the same code point? That, with a local `GraphemeBreakProperty` alias + a few newlines to break up large sections doesn't look too bad:

```c++
Vector<size_t> find_grapheme_segmentation_boundaries(Utf16View const& view)
{
#if ENABLE_UNICODE_DATA
    using GBP = GraphemeBreakProperty;
    Vector<size_t> boundaries;

    // https://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules
    if (view.length_in_code_points() == 0)
        return boundaries;

    auto has_any_gbp = [](u32 code_point, auto&&... properties) {
        return (code_point_has_grapheme_break_property(code_point, properties) || ...);
    };

    // GB1
    boundaries.append(0);

    if (view.length_in_code_points() > 1) {
        auto it = view.begin();
        auto code_point = *it;
        u32 next_code_point;
        auto current_ri_chain = 0;
        auto in_emoji_sequence = false;

        for (++it; it != view.end(); ++it, code_point = next_code_point) {
            next_code_point = *it;

            auto code_point_is_cr = has_any_gbp(code_point, GBP::CR);
            auto next_code_point_is_lf = has_any_gbp(next_code_point, GBP::LF);

            // GB3
            if (code_point_is_cr && next_code_point_is_lf)
                continue;
            // GB4, GB5
            if (code_point_is_cr || next_code_point_is_lf || has_any_gbp(next_code_point, GBP::CR, GBP::Control) || has_any_gbp(code_point, GBP::LF, GBP::Control)) {
                boundaries.append(view.code_unit_offset_of(it));
                continue;
            }

            auto next_code_point_is_v = has_any_gbp(next_code_point, GBP::V);
            auto next_code_point_is_t = has_any_gbp(next_code_point, GBP::T);

            // GB6
            if (has_any_gbp(code_point, GBP::L) && (next_code_point_is_v || has_any_gbp(next_code_point, GBP::L, GBP::LV, GBP::LVT)))
                continue;
            // GB7
            if (has_any_gbp(code_point, GBP::LV, GBP::V) && (next_code_point_is_v || next_code_point_is_t))
                continue;
            // GB8
            if (has_any_gbp(code_point, GBP::LVT, GBP::T) && next_code_point_is_t)
                continue;

            auto code_point_is_zwj = has_any_gbp(code_point, GBP::ZWJ);
            if (!in_emoji_sequence && code_point_has_property(code_point, Property::Extended_Pictographic))
                in_emoji_sequence = true;
            else if (in_emoji_sequence && !has_any_gbp(code_point, GBP::Extend) && !code_point_is_zwj)
                in_emoji_sequence = false;

            // GB9
            if (has_any_gbp(next_code_point, GBP::Extend, GBP::ZWJ))
                continue;
            // GB9a
            if (has_any_gbp(next_code_point, GBP::SpacingMark))
                continue;
            // GB9b
            if (has_any_gbp(code_point, GBP::Prepend))
                continue;

            // GB11
            if (in_emoji_sequence && code_point_is_zwj && code_point_has_property(next_code_point, Property::Extended_Pictographic))
                continue;

            auto code_point_is_ri = has_any_gbp(code_point, GBP::Regional_Indicator);
            current_ri_chain = code_point_is_ri ? current_ri_chain + 1 : 0;

            // GB12, GB13
            if (code_point_is_ri && has_any_gbp(next_code_point, GBP::Regional_Indicator) && current_ri_chain % 2 == 1)
                continue;

            // GB999
            boundaries.append(view.code_unit_offset_of(it));
        }
    }

    // GB2
    boundaries.append(view.length_in_code_units());
    return boundaries;
#else
    return {};
#endif
}
```

File: Userland/Libraries/LibJS/Runtime/Intl/Segmenter.cpp [hunk 3]
```diff
@@ -40,4 +43,73 @@ StringView Segmenter::segmenter_granularity_string() const
...
+double find_boundary(Segmenter const& segmenter, Utf16View const& string, double start_index, Direction direction, Optional<Vector<size_t>>& boundaries_cache)
+{
+    // 1. Let locale be segmenter.[[Locale]].
+    auto const& locale = segmenter.locale();
+
+    // 2. Let granularity be segmenter.[[SegmenterGranularity]].
+    auto granularity = segmenter.segmenter_granularity();
+
+    // 3. Let len be the length of string.
+    auto len = string.length_in_code_units();
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/Segmenter.cpp:105]: Nit, I kinda prefer real words like `length` when the spec says `len`

File: Userland/Libraries/LibJS/Runtime/Intl/Segments.h [hunk 4]
```diff
@@ -25,11 +25,15 @@ class Segments final : public Object {
 
     Utf16View segments_string() const { return m_segments_string.view(); }
 
+    Optional<Vector<size_t>>& boundaries_cache() const { return m_boundaries_cache; }
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Runtime/Intl/Segments.h:28]: Since the method is const, doesn't this need to return a `const &`?

Comment (trflynn89) [reply Userland/Libraries/LibJS/Runtime/Intl/Segments.h:28]: Oh, it's mutable, nvm

File: Userland/Libraries/LibUnicode/CharacterTypes.cpp (hunk 2 above)
Comment (IdanHo) [reply Userland/Libraries/LibUnicode/CharacterTypes.cpp:361]: That looks much nicer, thank you! :)

Comment (IdanHo) [reply Userland/Libraries/LibUnicode/CharacterTypes.cpp:361]: Updated the rest as well.

**Review by trflynn89 — APPROVED** [review]

**Comment by linusg** [comment]
Ref #9633

**Pull request merged**